File: plugins/DungeonForge/src/main/java/nl/riddernix/dungeonforge/party/PartyListener.java
package nl.riddernix.dungeonforge.party;

import nl.riddernix.dungeonforge.DungeonForgePlugin;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;

/** Retains party-instance occupancy across disconnects and restores rejoiners. */
public final class PartyListener implements Listener {

    private final DungeonForgePlugin plugin;

    public PartyListener(DungeonForgePlugin plugin) {
        this.plugin = plugin;
    }

    @EventHandler
    public void onQuit(PlayerQuitEvent event) {
        plugin.rooms().forget(event.getPlayer());
        plugin.parties().rememberDisconnect(event.getPlayer());
    }

    @EventHandler
    public void onJoin(PlayerJoinEvent event) {
        if (plugin.parties().takeCompletionReturn(event.getPlayer().getUniqueId()).map(location -> {
            plugin.getServer().getScheduler().runTask(plugin, () -> event.getPlayer().teleport(location));
            return true;
        }).orElse(false)) return;
        plugin.parties().reconnectLocation(event.getPlayer()).ifPresent(location ->
                plugin.getServer().getScheduler().runTask(plugin, () -> {
                    event.getPlayer().teleport(location);
                    plugin.rooms().refresh(event.getPlayer());
                }));
    }
}

File: plugins/DungeonPlugin/src/main/kotlin/nl/riddernix/dungeonplugin/party/PartyListener.kt
package nl.riddernix.dungeonplugin.party

import nl.riddernix.dungeonplugin.DungeonPlugin
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent

/** Retains party-instance occupancy across disconnects and restores rejoiners. */
class PartyListener(private val plugin: DungeonPlugin) : Listener {

    @EventHandler
    fun onQuit(event: PlayerQuitEvent) {
        plugin.rooms.forget(event.player)
        plugin.parties.rememberDisconnect(event.player)
    }

    @EventHandler
    fun onJoin(event: PlayerJoinEvent) {
        val completionReturn = plugin.parties.takeCompletionReturn(event.player.uniqueId)
        if (completionReturn != null) {
            plugin.server.scheduler.runTask(plugin, Runnable { event.player.teleport(completionReturn) })
            return
        }
        plugin.parties.reconnectLocation(event.player)?.let { location ->
            plugin.server.scheduler.runTask(plugin, Runnable {
                event.player.teleport(location)
                plugin.rooms.refresh(event.player)
            })
        }
    }
}
